samples/compiler_plugins/xnnpack/src/xnnpack/Conversion/StablehloToXnnpack.cpp

```diff
@@ -26,6 +28,58 @@ static llvm::cl::opt<std::string> patternFileName(
     "xnnpack-pattern-file", llvm::cl::desc("file for pattern bytecode"),
     llvm::cl::init(""));
 
+/// Custom constraint invoked from PDL.
+static LogicalResult checkI4RankedTensorType(PatternRewriter &rewriter,
+                                             Value value) {
+  if (auto ty = dyn_cast<RankedTensorType>(value.getType())) {
+    return success(ty.getElementType().isInteger(4));
+  }
+  return failure();
```

Comment:
nit: use `rewriter.notifyMatchFailure(...)` with an error message to make it easier to know why the pattern failed to match. Same comment for other `failure`s below